public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/PyroscopeAppIntegration.tsx

```diff
@@ -0,0 +1,133 @@
+
+/** A subset of the app settings that are relevant for this integration */
+type PyroscopeAppSettings = {
+  enabled: boolean;
+  jsonData: {
+    backendUrl: string;
+    basicAuthUser: string;
+  }
+}
+
+/** A subset of the datasource settings that are relevant for this integration */
+type PyroscopeDatasourceSettings = {
+  url: string;
+  basicAuthUser: string | number; // This might be configured as a number, so we will want to ensure compare for string-equivalence
+}
+
+export function PyroscopeAppIntegration(props: Props) {
+
+  const [appPlugin, setAppPlugin] = useState<PyroscopeAppSettings>()
+  const [datasource, setDatasource] = useState<PyroscopeDatasourceSettings>()
+
+  const {datasource:{uid: datasourceUid}, query, range} = props;
+
+  useEffect(()=>{
+
+    if (pyroscopeAppSettings == null) {
+      pyroscopeAppSettings = getBackendSrv().get<PyroscopeAppSettings>(`/api/plugins/${PYROSCOPE_APP_ID}/settings`)
+    }
+
+    pyroscopeAppSettings.then(setAppPlugin);
+    
+  }, [])
+
+  useEffect(()=>{
+    let datasourceSettings = pyroscopeDatasourceSettingsByUid[datasourceUid];
+
+    if (datasourceSettings == null) {
+      datasourceSettings = getBackendSrv().get<PyroscopeDatasourceSettings>(`/api/datasources/uid/${datasourceUid}`)
+      pyroscopeDatasourceSettingsByUid[datasourceUid] = datasourceSettings;
+    }
+
+    datasourceSettings.then(setDatasource);
+
+  }, [datasourceUid])
+
+  const queryParam = useMemo(()=>{
+
+    return generateQueryParams(query, range);
+  }, [query, range])
+
+  const profilesAppLink = useMemo(()=>isPyroscopeDatasourceCompatibleWithPlugin(datasource, appPlugin) &&
+    <LinkButton variant='secondary' icon='external-link-alt' tooltip={'Open query in Profiles App'} target='_blank' href={`/a/${PYROSCOPE_APP_ID}/single?${queryParam}`}>Profiles App</LinkButton>
+  , [appPlugin, datasource, queryParam])
+
+  return profilesAppLink || null;
+}
+
+export function isPyroscopeDatasourceCompatibleWithPlugin(datasource?: PyroscopeDatasourceSettings, appPlugin?: PyroscopeAppSettings) {
+  if (!appPlugin || !datasource) {
+    return false;
+  }
+
+  return (
+    datasource.url === appPlugin.jsonData.backendUrl
+    &&
+    String(datasource.basicAuthUser) === String(appPlugin.jsonData.basicAuthUser)
+  )
+}
+
+
+function stringifyRawTimeRangePart(rawTimeRangePart: DateTime | string) {
+  if (typeof rawTimeRangePart === 'string') {
+    return rawTimeRangePart;
+  }
+
+  // The `unix` result as a string is compatible with Pyroscope's range part format
+  return Math.round(rawTimeRangePart.unix()).toString();
+}
+
+export function translateGrafanaTimeRangeToPyroscope(timeRange: TimeRange) {
+  const from = stringifyRawTimeRangePart(timeRange.raw.from);
+  const until = stringifyRawTimeRangePart(timeRange.raw.to);
+
+  return { from, until };
+}
+
+export function generateQueryParams(query?: Query, range?: TimeRange) {
```

Comment:
Query should always be defined based on the `QueryEditorProps` (and transitively in this component) so not sure this needs to be optional.